Salesforce migration

djagoda881 opened the pull request:
<!-- Thanks for contributing to viadot! 🙏-->

## Summary
<!-- A sentence summarizing the PR -->
The Salesforce source was moved from viadot version 1. The source was moved along with the unit tests. The structure of the source has undergone the following changes compared to viadot 1:

- Added class attribute config_key
- Updated how secrets are retrieved. After the change, it looks the same as in other sources, e.g. exchange_rates.

## Importance
<!-- Why is this PR important? -->
Marge It is needed to use source in prefect-viadot container

## Checklist
<!-- PRs will not be reviewed unless these boxes are checked -->

This PR:

- [x ] adds new tests (if appropriate)
- [ x] updates docstrings for any new functions or function arguments (if appropriate)
- [ x] updates `CHANGELOG.md` with a summary of the changes

Cuppingan commented:
Advanced Communities optimized the way non-profits interact with donors, volunteers, members, and constituents. They can help your non-profit organization make a measurable impact on the world around us. Give your customers the choice to pay online or offline with secure debit and credit card payments. Track all transactions easily with Stripe and native support of FinDock payments and donations integrations. Check this Salesforce e-commerce solution https://advancedcommunities.com/products/ac-ecommerce-for-non-profits/

djagoda881 commented:
I made some minor corrections, the code is now ready for review

Trymzet commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,192 @@
+
+class Salesforce(Source):
+    """
+    A class for pulling data from theSalesforce.
```
typo

djagoda881 replied:
[📝 Updated main class docstring](https://github.com/dyvenia/viadot/pull/629/commits/bf79a66a06f490d315e823857fb35708621e2114)

Trymzet commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,192 @@
+    A class for pulling data from theSalesforce.
+
+    Args:
+        domain (str): domain of a connection; defaults to 'test' (sandbox). Can be added only if built-in username/password/security token is provided.
+        client_id (str): client id to keep the track of API calls.
+        credentials (dict): credentials to connect with. If not provided, will read from local config file.
+        env (Literal): environment information, provides information about credential and connection configuration; defaults to 'DEV'.
+        config_key (str, optional): The key in the viadot config holding relevant credentials. Defaults to None.
```
Start with capital letter

djagoda881 replied:
[📝 Updated main class docstring](https://github.com/dyvenia/viadot/pull/629/commits/bf79a66a06f490d315e823857fb35708621e2114)

Trymzet commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,192 @@
+        super().__init__(*args, credentials=credentials, **kwargs)
+
+        if env.upper() == "DEV":
+            self.salesforce = SF(
+                username=self.credentials.get("username"),
+                password=self.credentials.get("password"),
+                security_token=self.credentials.get("token"),
+                domain=domain,
+                client_id=client_id,
+            )
+        elif env.upper() == "QA":
+            self.salesforce = SF(
+                username=self.credentials.get("username"),
+                password=self.credentials.get("password"),
+                security_token=self.credentials.get("token"),
+                domain=domain,
+                client_id=client_id,
+            )
+        elif env.upper() == "PROD":
+            self.salesforce = SF(
+                username=self.credentials.get("username"),
+                password=self.credentials.get("password"),
+                security_token=self.credentials.get("token"),
+                domain=domain,
+                client_id=client_id,
+            )
```
This looks exactly the same, why the if/else? BTW are you sure this is correct, I think there was a difference for some envs in Salesforce reg. the parameters that needed to be provided.

Trymzet commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,192 @@
+        external_id: str = None,
+        raise_on_error: bool = False,
+    ) -> None:
+
```
Can you add a docstring here pls?

Trymzet commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,192 @@
+
+        for record in records_cp:
+            response = 0
+            if external_id:
+                if record[external_id] is None:
+                    continue
```
This begs a comment

djagoda881 replied:
[🎨 Improved retrieving externalID](https://github.com/dyvenia/viadot/pull/629/commits/97e6dbb9a79abf04b0787c07e8f44830a654fb8f)

Remove test code:

[⏪ Removed tests code](https://github.com/dyvenia/viadot/pull/629/commits/5625cee1776787cbea2faf7ab6a79067fb7491e5)

Trymzet commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,192 @@
+                else:
+                    self.logger.warning(msg)
+
+            codes = {200: "updated", 201: "created", 204: "updated"}
```
Can you rename to `valid_response_codes` pls

djagoda881 replied:
[🎨 Changed name of 'code' var to 'valid_response_code'](https://github.com/dyvenia/viadot/pull/629/commits/f976ecd5af528363e8ee21306e2fbe4c824d91be)

Trymzet commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,192 @@
+                merge_key = record.pop("Id")
+
+            try:
+                response = table_to_upsert.upsert(data=record, record_id=merge_key)
```
Can you rename to `response_code` pls

djagoda881 replied:
[🎨 Changed name of 'response' var to 'response_code'](https://github.com/dyvenia/viadot/pull/629/commits/f5deb9aba7c1912a86f9d4087fab1ebe475a3e0b)

Trymzet commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,192 @@
+        batch_size: int = 10000,
+        raise_on_error: bool = False,
+    ) -> None:
+
```
Add docstring pls

Trymzet commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,192 @@
+                f"Passed DataFrame does not contain column '{external_id}'."
+            )
+        records = df.to_dict("records")
+        response = 0
```
That doesn't seem right as we iterate through `response` on line 154, so it can't be an integer? Also, anyway this variable is overwritten two lines below, I don't understand the purpose of this line.

djagoda881 replied:
[🔥 Removed response_code var declaration](https://github.com/dyvenia/viadot/pull/629/commits/ddee24f7af437d9fbf215a060ce8e8f6cbc04fc8)

Trymzet commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,192 @@
+
+    def download(
+        self, query: str = None, table: str = None, columns: List[str] = None
+    ) -> List[OrderedDict]:
```
Add docstring pls

djagoda881 replied:
[📝 Added docstring to Slesforce class functions](https://github.com/dyvenia/viadot/pull/629/commits/a3fea0e7499287c554de2babe2f6367de9a8f058)

Trymzet commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,192 @@
+                columns_str = "FIELDS(STANDARD)"
+            query = f"SELECT {columns_str} FROM {table}"
+        records = self.salesforce.query(query).get("records")
+        # Take trash out.
```
I guess we can add sth better here, like "Remove metadata from the data"

djagoda881 replied:
[💡 Updated comment in download fuction ](https://github.com/dyvenia/viadot/pull/629/commits/3aaad0644e57431a62871c08bc0f326d963fda7c)

Trymzet commented on tests/unit/test_salesforce.py:
```diff
@@ -0,0 +1,101 @@
+
+@pytest.fixture(scope="session")
+def salesforce():
+    s = Salesforce(config_key="sales_force_dev")
```
Salesforce is written together, so should be `salesforce_dev`

djagoda881 replied:
[🔐 Updated config_key name after git revert](https://github.com/dyvenia/viadot/pull/629/commits/a8fa33590c9a266a5ac5f9099b6ffed78e700b08)

Trymzet commented on tests/unit/test_salesforce.py:
```diff
@@ -0,0 +1,101 @@
+        "LastName": ["LastName"],
+    }
+    df_restored = pd.DataFrame(data=data_restored)
+    salesforce.upsert(df=df_restored, table=TABLE_TO_UPSERT)
```
You should not use a function you're about to test; also the teardown should be cleaning up resources, not creating them

djagoda881 replied:
[✅ Improved cleaning after tests](https://github.com/dyvenia/viadot/pull/629/commits/b408db7f60c099bc43421eaf070158d13e2a1f8d)

Trymzet replied:
The second point remains, I think it may be confusing because you're mixing 2 things together in these fixtures, you're creating a test df but also randomly upserting records. It's actually not easy to understand the tests since you already upsert records in the fixture, I don't think this should be happening? Perhaps you meant to **insert** some records in order to validate  that they are correctly updated in the `Salesforce.upsert()` method? But in that case, I don't think this needs to be in a fixture, you can do this (note also there is a cleanup at the end, pls also add it so the tests do not permanently create/modify any records).
```python
TEST_DF_EXTERNAL = pd.DataFrame({"LastName": [TEST_LAST_NAME], "SAPContactId__c": ["111"]})

def test_upsert(salesforce):
    # Setup.
    sf = salesforce.salesforce
    test_record = TEST_DF_EXTERNAL.to_records()  # not sure if this is the data structure required by insert(), pls check
    sf.Contact.insert(test_record)
     
    # Test.
    salesforce.upsert(
        df=TEST_DF_EXTERNAL, table=TABLE_TO_UPSERT, external_id="SAPContactId__c"
    )

    result = sf.query(
        f"SELECT ID, LastName FROM {TABLE_TO_UPSERT} WHERE ID='{ID_TO_UPSERT}'"
    )
    assert result["records"][0]["LastName"] == TEST_LAST_NAME

    # Cleanup
    sf.delete(test_record)
```

Trymzet commented on tests/unit/test_salesforce.py:
```diff
@@ -0,0 +1,101 @@
+    salesforce.upsert(df=df_restored, table=TABLE_TO_UPSERT)
+
+
+def test_upsert_empty(salesforce):
+    try:
```
Can you add a docstring here explaining this tests that an exception is not raised?

djagoda881 replied:
I will add, but I think I will remove this test because it doesn't make much sense. Because in other upserts there is a try exception block that will catch the error

djagoda881 replied:
[✅ 🔥 Removed test test_upsert_empty](https://github.com/dyvenia/viadot/pull/629/commits/ab1d1af708b47241f6ebaefa05311edb616c1bcb)

Trymzet commented on tests/unit/test_salesforce.py:
```diff
@@ -0,0 +1,101 @@
+
+
+def test_upsert_external_id_correct(salesforce, test_df_external):
+    try:
+        salesforce.upsert(
+            df=test_df_external, table=TABLE_TO_UPSERT, external_id="SAPContactId__c"
+        )
+    except Exception as exception:
+        assert False, exception
```
don't retest what another test has already tested

djagoda881 replied:
This test checks the upsert based on external_ID. Other tests perform an upsert based on the row ID

djagoda881 replied:
I corrected the structure of this test
[✅ 🎨Improved structure of upsert test](https://github.com/dyvenia/viadot/pull/629/commits/de73e4ca66fa277fb67d44c7cd5221f0c9eacc00)

Trymzet commented on tests/unit/test_salesforce.py:
```diff
@@ -0,0 +1,101 @@
+        assert False, exception
+
+
+def test_upsert_external_id_correct(salesforce, test_df_external):
```
It seems like this only tests the insert part, and not the update part?

djagoda881 replied:
This test only performs an update. For some reason I couldn't insert on records

djagoda881 replied:
I corrected the structure of this test
[✅ 🎨Improved structure of upsert test](https://github.com/dyvenia/viadot/pull/629/commits/de73e4ca66fa277fb67d44c7cd5221f0c9eacc00)

Trymzet replied:
You're saying that this function doesn't work? Then this test should be failing. The upsert test needs to test two things:
- new records are inserted
- existing records are updated

This should also work on multiple records (`df` can have eg. 100 rows), not on a single row as you suggerst [here](https://github.com/dyvenia/viadot/blob/a3fea0e7499287c554de2babe2f6367de9a8f058/viadot/sources/salesforce.py#L81).

djagoda881 replied:
I renamed the column and added a test inserting two rows using the upsert method.

[🎨Rename external_id var to external_id_column](https://github.com/dyvenia/viadot/pull/629/commits/b6e311e43d8a83f64c00a58fcc17fc4c52db9140)

[✅ Added inserting two rows](https://github.com/dyvenia/viadot/pull/629/commits/c8bdf4e81fad33033ce2bf5fee2f7aa48e5355cb)

Trymzet commented on tests/unit/test_salesforce.py:
```diff
@@ -0,0 +1,101 @@
+
+
+def test_download_no_query(salesforce):
+    ordered_dict = salesforce.download(table=TABLE_TO_DOWNLOAD)
```
Looking at the typing the type of this should actually be a list of ordered dicts, but anyway it's usually not good to name variables by their type, why not just call this sth like `records`?

djagoda881 replied:
[✅ 🎨 Changed name of variable](https://github.com/dyvenia/viadot/pull/629/commits/3e2ef0856cf130df631664474368f0fd15210bc7)

Trymzet commented on tests/unit/test_salesforce.py:
```diff
@@ -0,0 +1,101 @@
+
+
+def test_to_df(salesforce):
+    df = salesforce.to_df(table=TABLE_TO_DOWNLOAD)
+    assert df.empty == False
```
Can you pls also add an equality check?

djagoda881 replied:
The table is very large, is checking its dimensions sufficient?
[✅🎨Improved structure of test_to_df](https://github.com/dyvenia/viadot/pull/629/commits/61bf1610b01be0ae459e7578558c6df73edf96db)

Trymzet commented on tests/unit/test_salesforce.py:
```diff
@@ -0,0 +1,101 @@
+    assert df.empty == False
+
+
+def test_upsert(salesforce, test_df_data):
+    salesforce.upsert(df=test_df_data, table=TABLE_TO_UPSERT)
```
Looks unfinished

djagoda881 replied:
[✅ 🎨 Improved structure of test_upsert](https://github.com/dyvenia/viadot/pull/629/commits/47e177ecfe466d1a64c45b3db34e765f2b99cfc1)

Trymzet reviewed (changes requested):
Some minor adjustments and missing docstrings

Trymzet commented on tests/unit/test_salesforce.py:
```diff
@@ -4,7 +4,7 @@
 TABLE_TO_DOWNLOAD = "Account"
 TABLE_TO_UPSERT = "Contact"
-TEST_LAST_NAME = "prefect-viadot-test"
+TEST_LAST_NAME = "viadot-test"
```
👍

Trymzet commented on tests/unit/test_salesforce.py:
```diff
@@ -77,7 +77,10 @@ def test_download_with_query(salesforce):
 
 def test_to_df(salesforce):
     df = salesforce.to_df(table=TABLE_TO_DOWNLOAD)
+    print(len(df.values))
```
Remove print

djagoda881 replied:
[✅ Improved test_to_df](https://github.com/dyvenia/viadot/pull/629/commits/5adfbe2b9b0de1c1cfc632b19bfc3849512b3f98)

Trymzet commented on tests/unit/test_salesforce.py:
```diff
@@ -77,7 +77,10 @@ def test_download_with_query(salesforce):
+    print(len(df.values))
     assert df.empty == False
+    assert len(df.columns) == 98
+    assert len(df.values) >= 1000
```
Seems like a lot of rows for testing purposes, is it possible to limit this? I think we don't need more than say, 3 rows?

djagoda881 replied:
[✅ Improved test_to_df](https://github.com/dyvenia/viadot/pull/629/commits/5adfbe2b9b0de1c1cfc632b19bfc3849512b3f98)

Trymzet commented on viadot/sources/salesforce.py:
```diff
@@ -173,6 +193,18 @@ def bulk_upsert(
         self, query: str = None, table: str = None, columns: List[str] = None
     ) -> List[OrderedDict]:
+        """
+        Dowload all data from the indicated table or the result of the specified query.
```
typo

djagoda881 replied:
[✅📝Improved docstings in Salesforce class](https://github.com/dyvenia/viadot/pull/629/commits/570b5797c61f1e1ca58ec24b03d4d67b96d66706)

Trymzet reviewed (changes requested):
Added a couple of comments

djagoda881 commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,232 @@
+        raise_on_error: bool = False,
+    ) -> None:
+        """
+        Performs upsert operations on the selected row in the table.
```
[✅📝Improved docstings in Salesforce class](https://github.com/dyvenia/viadot/pull/629/commits/570b5797c61f1e1ca58ec24b03d4d67b96d66706)

djagoda881 commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,232 @@
+        Performs upsert operations on the selected row in the table.
+
+        Args:
+            df (pd.DataFrame): The DataFrame to upsert. Only a single row can be upserted with this function.
```
[✅📝Improved docstings in Salesforce class](https://github.com/dyvenia/viadot/pull/629/commits/570b5797c61f1e1ca58ec24b03d4d67b96d66706)

djagoda881 commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,232 @@
+        raise_on_error: bool = False,
+    ) -> None:
+        """
+        Performs upsert operations on multiple rows in a table.
```
[✅📝Improved docstings in Salesforce class](https://github.com/dyvenia/viadot/pull/629/commits/570b5797c61f1e1ca58ec24b03d4d67b96d66706)

djagoda881 commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,232 @@
+        domain: str = "test",
+        client_id: str = "viadot",
+        env: Literal["DEV", "QA", "PROD"] = "DEV",
+        credentials: Dict[str, Any] = None,
```
Created https://github.com/dyvenia/viadot/issues/651

djagoda881 commented:
I decided to change the test structure by looking at the advice from 1:1 on Thursday. 

- I have added a bulk upsert test

- Each test is self-contained, i.e. rows are created and deleted in each individual test

- At the end of the session, it is checked whether any row remains in the database and is deleted if found  

- In an upsert/bulk_upsert test the insert is performed first and then the update is checked in the same test

- I added a `get_tested_records` function which returns the result of the query. I added it because very often I had to call the same qyery to validate the tests


[✅ Added bulk upsert test and refeactored tests](https://github.com/dyvenia/viadot/pull/629/commits/990775a64baf15b9f9fb5eb1b7d19dbc24a53dba)

Trymzet commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,238 @@
+        Args:
+            query (str, optional): Query for download the specific data. Defaults to None.
+            table (str, optional): Table name. Defaults to None.
+            columns (List[str], optional): List of columns which are needed,
```
"List of columns which are needed, requires table argument"-> "list of required columns; requires `table` to be specified" (appears 2x)

djagoda881 replied:
[📝 Updated docstrings in Salesforce source](https://github.com/dyvenia/viadot/pull/629/commits/03d07c4da6508d2b436bf0a4bb09e4f7c3a3db54)

Trymzet commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,238 @@
+        columns: List[str] = None,
+    ) -> pd.DataFrame:
+        """
+        Downloads the indicated data and returns the Dataframe.
```
Downloads the indicated data returns it as a pandas `DataFrame`

djagoda881 replied:
[📝 Updated docstrings in Salesforce source](https://github.com/dyvenia/viadot/pull/629/commits/03d07c4da6508d2b436bf0a4bb09e4f7c3a3db54)

djagoda881 commented on viadot/sources/salesforce.py:
```diff
@@ -137,10 +138,13 @@ def bulk_upsert(
 
         Args:
-            df (pd.DataFrame): The DataFrame to upsert.
+            df (pd.DataFrame): Dataframe containing the rows to Bulk upsert.
```
[📝 Updated docstrings in Salesforce source](https://github.com/dyvenia/viadot/pull/629/commits/03d07c4da6508d2b436bf0a4bb09e4f7c3a3db54)

djagoda881 commented on viadot/sources/salesforce.py:
```diff
@@ -70,10 +70,11 @@ def upsert(
         """
-        Performs upsert operations on the selected row in the table.
+        Upsert the DataFrame to Salesforce. The upsert is performed on a single record at a time.
+        Using an upsert operation gives you more control over logging and error handling than using Bulk upsert.
```
[📝 Updated docstrings in Salesforce source](https://github.com/dyvenia/viadot/pull/629/commits/03d07c4da6508d2b436bf0a4bb09e4f7c3a3db54)

djagoda881 commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,238 @@
+
+    Args:
+        domain (str, optional): Domain of a connection. Defaults to 'test' (sandbox).
+            Can be added only if built-in username/password/security token is provided.
```
[📝 Updated docstrings in Salesforce source](https://github.com/dyvenia/viadot/pull/629/commits/03d07c4da6508d2b436bf0a4bb09e4f7c3a3db54)

djagoda881 commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,238 @@
+    Args:
+        domain (str, optional): Domain of a connection. Defaults to 'test' (sandbox).
+            Can be added only if built-in username/password/security token is provided.
+        client_id (str, optional): Client id to keep the track of API calls.
```
[📝 Updated docstrings in Salesforce source](https://github.com/dyvenia/viadot/pull/629/commits/03d07c4da6508d2b436bf0a4bb09e4f7c3a3db54)

djagoda881 commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,238 @@
+        Using an upsert operation gives you more control over logging and error handling than using Bulk upsert.
+
+        Args:
+            df (pd.DataFrame): Dataframe containing the rows to upsert.
```
[📝 Updated docstrings in Salesforce source](https://github.com/dyvenia/viadot/pull/629/commits/03d07c4da6508d2b436bf0a4bb09e4f7c3a3db54)

djagoda881 commented on viadot/sources/salesforce.py:
```diff
@@ -0,0 +1,238 @@
+        Download all data from the indicated table or the result of the specified query.
+
+        Args:
+            query (str, optional): Query for download the specific data. Defaults to None.
```
[📝 Updated docstrings in Salesforce source](https://github.com/dyvenia/viadot/pull/629/commits/03d07c4da6508d2b436bf0a4bb09e4f7c3a3db54)

djagoda881 commented on tests/unit/test_salesforce.py:
```diff
@@ -28,7 +28,7 @@ def test_row_creation(salesforce):
 
 
-def test_upsert_external_id_correct(salesforce, test_row_creation):
+def test_upsert_external_id_column(salesforce, test_row_creation):
```
This is the name of the test after the latest changes

djagoda881 commented on tests/unit/test_salesforce.py:
```diff
@@ -76,14 +78,39 @@ def test_upsert_row_id(salesforce, test_row_creation):
+    try:
+        salesforce.upsert(df=df, table="Contact", external_id_column="SAPContactId__c")
+    except Exception as exception:
+        raise exception
```
I have removed the try except block from the tests. Inside the function in upsert and bulk upsert the commands are executed in the try except block so if they don't work correctly an error will appear

djagoda881 commented on tests/unit/test_salesforce.py:
```diff
@@ -0,0 +1,184 @@
+    sf.Contact.delete(updated_row[0]["Id"])
+
+
+def test_upsert(salesforce):
```
In `test_upsert `it performs upsert operations based on the `external_id`. In this test, performs an `upsert `operation based on the `Id` of the row (the `Id `is allocated when the new row is created )

djagoda881 commented:
The DE team has created a new version of this PR.

djagoda881 commented:
New PR: https://github.com/dyvenia/viadot/pull/708
